### PR TITLE
Optimize away some physical equality

### PR DESCRIPTION
**Pull request by chambart**
Physical equality on structurally different values is always false. This patch allows to handle some cases. The need arise when evaluating some functions that use it to avoid some computation. For instance in `Set.add` test if the added value is already present to avoid rebuilding the same set.

``` OCaml
    let rec add x = function
        Empty -> Node(Empty, x, Empty, 1)
      | Node(l, v, r, _) as t ->
          let c = Ord.compare x v in
          if c = 0 then t else
          if c < 0 then
            let ll = add x l in
            if l == ll then t else bal ll v r
          else
            let rr = add x r in
            if r == rr then t else bal l v rr
```

This change allows some of those uses to be eliminated.

### CODE REVIEW

File: middle_end/simplify_primitives.ml [hunk 1]
```diff
@@ -37,6 +37,39 @@ let phys_equal (approxs:A.t list) =
     | Some (s1, Some f1), Some (s2, Some f2) -> Symbol.equal s1 s2 && f1 = f2
     | _ -> false
 
+let is_known_to_be_some_kind_of_int (arg:A.descr) =
+  match arg with
+  | Value_int _ | Value_char _ | Value_constptr _ -> true
+  | _ -> false
```
Comment (mshinwell) [inline middle_end/simplify_primitives.ml]: Please make both of these matches exhaustive.

File: middle_end/simplify_primitives.ml [hunk 2]
```diff
@@ -37,6 +37,39 @@ let phys_equal (approxs:A.t list) =
...
+let is_known_to_be_some_kind_of_block (arg:A.descr) =
+  match arg with
+  | Value_block _ | Value_float _ | Value_float_array _ | A.Value_boxed_int _
+  | Value_closure _ | Value_string _ -> true
+  | _ -> false
+
+let rec structurally_different (arg1:A.t) (arg2:A.t) =
+  match arg1.descr, arg2.descr with
+  | Value_block (tag1, fields1), Value_block (tag2, fields2) ->
+      not (Tag.equal tag1 tag2) ||
```
Comment (mshinwell) [inline middle_end/simplify_primitives.ml]: Indentation needs fixing for these two stanzas (also on line 69 too).
I would also recommend following the GNU convention of putting the `||` at the start of each line of a multi-line conditional; I think it's easier to follow.

**Review by mshinwell — CHANGES_REQUESTED** [review]

**Comment by mshinwell** [comment]
@chambart Ping

File: testsuite/tests/asmcomp/is_static_flambda.ml [hunk 3]
```diff
@@ -113,3 +113,88 @@ let f x =
...
+   known constant since the function is never inlined (hence this
+   code is never at toplevel) *)
+
+let () =
+  let f () =
+    let v = (1, 2) in
+    (* eq is supposed to be considered always true since v is a
+       constant, hence aliased to a symbol.
+       It is not yet optimized away if it is not constant *)
+    let eq = v == v in
```
Comment (mshinwell) [inline testsuite/tests/asmcomp/is_static_flambda.ml:129]: These tests might as well check that `eq` is as expected, no?

**Comment by mshinwell** [comment]
I made one comment, but OK otherwise.  We need to wait for @lpw25 to review too.

**Comment by mshinwell** [comment]
Ah, this should have a Changes entry too, I think.

**Review by mshinwell — APPROVED** [review]

**Comment by mshinwell** [comment]
@chambart Please add a Changes entry and merge to trunk, thanks.

**Comment by alainfrisch** [comment]
Ping @chambart

File: testsuite/tests/asmcomp/is_static_flambda.ml (hunk 3 above)
Comment (mshinwell) [reply testsuite/tests/asmcomp/is_static_flambda.ml:129]: Actually I think this is fine as-is.

**Comment by mshinwell** [comment]
I've added the Changes entry and merged with current trunk.  I will merge into trunk assuming CI passes.

**Pull request merged**